src/effects/VST3/VST3UIValidator.cpp

```diff
@@ -0,0 +1,205 @@
+      {
+         if(view->canResize() == Steinberg::kResultTrue)
+         {
+            ViewRect minSize {0, 0, parent->GetMinWidth(), parent->GetMinHeight()};
+            ViewRect maxSize {0, 0, 10000, 10000};
+            if(view->checkSizeConstraint(&minSize) != kResultOk)
+               minSize = defaultSize;
+
+            if(view->checkSizeConstraint(&maxSize) != kResultOk)
+               maxSize = defaultSize;
+
+            //No need to accommodate the off-by-one error with Steinberg::ViewRect
+            //as we do it with wxRect
+
+            if(defaultSize.getWidth() < minSize.getWidth())
+               defaultSize.right = defaultSize.left + minSize.getWidth();
+            if(defaultSize.getWidth() > maxSize.getWidth())
+               defaultSize.right = defaultSize.left + maxSize.getWidth();
+
+            if(defaultSize.getHeight() < minSize.getHeight())
+               defaultSize.bottom = defaultSize.top + minSize.getHeight();
+            if(defaultSize.getHeight() > maxSize.getHeight())
+               defaultSize.bottom = defaultSize.top + maxSize.getHeight();
+
+            parent->SetMinSize({minSize.getWidth(), minSize.getHeight()});
+            parent->SetMaxSize({maxSize.getWidth(), maxSize.getHeight()});
+         }
+         else
+         {
+            parent->SetMinSize({defaultSize.getWidth(), defaultSize.getHeight()});
+            parent->SetMaxSize(parent->GetMinSize());
+         }
+
+         parent->SetSize({defaultSize.getWidth(), defaultSize.getHeight()});
+      }
+
+#if __WXGTK__
+      mPlugView = view;
+      safenew internal::x11::SocketWindow(parent, wxID_ANY, view);
+      return true;
+#else
+
+      static const auto platformType =
+#  if __WXMAC__
+         kPlatformTypeNSView;
+#  elif __WXMSW__
+         kPlatformTypeHWND;
+#  else
+#     error "Platform not supported"
+#  endif
+      auto plugFrame = owned(safenew internal::PlugFrame { parent });
+      view->setFrame(plugFrame);
+      if(view->attached(parent->GetHandle(), platformType) != kResultOk)
+         return false;
+
+      mPlugView = view;
+      mPlugFrame = plugFrame;
+
+      return true;
+#endif
+   }
+   return false;
+}
+
+bool VST3UIValidator::IsGraphicalUI()
+{
+   return mPlugView != nullptr;
+}
+
+bool VST3UIValidator::ValidateUI()
+{
+   mAccess.ModifySettings([&](EffectSettings &settings){
+      if (mDuration != nullptr)
+         settings.extra.SetDuration(mDuration->GetValue());
+      mWrapper.FlushSettings(settings);
+      mWrapper.StoreSettings(settings);
+   });
+
+   return true;
+}
+
+void VST3UIValidator::OnClose()
+{
+   using namespace Steinberg;
+
+   mPlainUI = nullptr;
+   mParent = nullptr;
+   if(mPlugView)
+   {
+      mPlugView->setFrame(nullptr);
+      mPlugView->removed();
+      mPlugView = nullptr;
+      mPlugFrame = nullptr;
+   }
+}
+
+bool VST3UIValidator::UpdateUI()
+{
+   if(!mWrapper.IsActive())
+      mWrapper.FetchSettings(mAccess.Get());
```

Comment:
My understanding of these lines is that they update both the edit controller and the effect component after you change a preset.

The first, presumably, causes graphical updates.  The other affects the state of the processor, but this might be happening during processing.

Are we confident of the thread safety of this?

If not -- then `RealtimeProcessStart` in the Instance would have to complete the change of state of the effect component.  Would it be correct just to do a complete state change there, not using the pending changes interface?

Reply:
This also raises the question of the thread safety of the calls that may precede `UpdateUI` and those are `LoadFactoryDefaults`, `LoadFactoryPreset`, `ImportPresets`, `LoadUserPreset`.  Are they all safe to do while another thread is processing?

Reply:
>Are we confident of the thread safety of this?

It's safe to call setState/getState during processing, but this doesn't happen during processing `if(!isActive)`

>Would it be correct just to do a complete state change there, not using the pending changes interface?

I would object to that since `RealtimeProcessStart` is getting called for each block, I guess it's not how it supposed to work. Correct thing would be to use setState once (or when preset has changed), send changes via `IComponentHandler` and then save state at least once (in our case - when UI is closed)